bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -281,9 +449,177 @@ impl AuthenticationService {
+
+        tracing::info!("Registering this client for OIDC.");
+        let registration_response = oidc
+            .register_client(&authentication_server.issuer, oidc_metadata.clone(), None)
+            .await?;
+
+        let client_data = RegisteredClientData {
+            // The format of the credentials changes according to the client metadata that was sent.
+            // Public clients only get a client ID.
+            credentials: ClientCredentials::None {
+                client_id: registration_response.client_id.clone(),
+            },
+            metadata: oidc_metadata,
+        };
+        oidc.restore_registered_client(authentication_server, client_data).await;
+
+        tracing::info!("Persisting OIDC registration data.");
+        self.store_client_registration(oidc).await?;
+
+        Ok(())
+    }
+
+    /// Stores the current OIDC dynamic client registration so it can be re-used
+    /// if we ever log in via the same issuer again.
+    async fn store_client_registration(&self, oidc: &Oidc) -> Result<(), AuthenticationError> {
+        let issuer = oidc.issuer().ok_or(AuthenticationError::OidcNotSupported)?;
+        let client_id = oidc
+            .client_credentials()
+            .ok_or(AuthenticationError::OidcError {
+                message: String::from("Missing client registration."),
+            })?
+            .client_id()
+            .to_owned();
+
+        let registrations = OidcRegistrations::new(
+            &self.base_path,
+            self.oidc_configuration
+                .as_ref()
+                .map(|c| c.static_registrations.clone())
+                .unwrap_or_default(),
+        )?;
+        registrations.set_client_id(client_id, issuer.to_owned())?;
 
-        Ok(HomeserverLoginDetails { url, authentication_issuer, supports_password_login })
+        Ok(())
+    }
+
+    /// Attempts to load an existing OIDC dynamic client registration for the
+    /// currently configured issuer.
+    async fn load_client_registration(
+        &self,
+        oidc: &Oidc,
+        oidc_metadata: VerifiedClientMetadata,
+    ) -> bool {
+        let Some(issuer) = oidc.issuer() else {
+            return false;
+        };
+        let Some(registrations) = OidcRegistrations::new(
+            &self.base_path,
+            self.oidc_configuration
+                .as_ref()
+                .map(|c| c.static_registrations.clone())
+                .unwrap_or_default(),
+        )
+        .ok() else {
+            return false;
+        };
+        let Some(client_id) = registrations.client_id(issuer) else {
+            return false;
+        };
+
+        let client_data = RegisteredClientData {
+            credentials: ClientCredentials::None { client_id },
+            metadata: oidc_metadata,
+        };
+
+        let authentication_server = AuthenticationServerInfo::new(
+            issuer.to_owned(),
+            oidc.account_management_url().unwrap_or(None).map(|url| url.to_string()),
```

Comment:
You should  just use the `AuthenticationServerInfo` provided in `configure_oidc`.